Compiling on GCC 4.8

There are a couple issues that don't allow compiling, the most recent is [this](https://github.com/CrowCpp/Crow/commit/338b59f329cd02f5f7e0ed839d00e00acb8e2c37#diff-16f874deea4720184d66cd4a3b8f5c7e0206a02135db8bd7ded4f153e693b7fdR328), but also the changes introduced in #327.

Here's an error log:
```
In file included from crow.h:19:0,
                 from main.cpp:1:
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
crow/routing.h:584:88: error: no matching function for call to ‘wrapped_handler_call(crow::request&, crow::response&, const main()::__lambda53&)’
                   detail::wrapped_handler_call(req, res, f, std::forward<Args>(args)...);
                                                                                        ^
crow/routing.h:584:88: note: candidates are:
In file included from crow/routing.h:18:0,
                 from crow.h:19,
                 from main.cpp:1:
./crow/middleware.h:206:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:206:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = main()::__lambda53; Args = {}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:206:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:216:11: note: template<class F, class ... Args> typename std::enable_if<(crow::black_magic::is_callable<F, crow::request&, crow::response&, Args ...>::value && (! crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value))>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:216:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<(crow::black_magic::is_callable<F, crow::request&, crow::response&, Args ...>::value && (! crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value))>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = main()::__lambda53; Args = {}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:216:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:226:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:226:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = main()::__lambda53; Args = {}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:226:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:236:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::request, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:236:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::request, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = main()::__lambda53; Args = {}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:236:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:247:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, Args>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:247:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, Args>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = main()::__lambda53; Args = {}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]’
main.cpp:10:6:   required from here
./crow/middleware.h:247:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
In file included from crow.h:19:0,
                 from main.cpp:1:
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
crow/routing.h:584:88: error: no matching function for call to ‘wrapped_handler_call(crow::request&, crow::response&, const crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50&, std::basic_string<char>)’
                   detail::wrapped_handler_call(req, res, f, std::forward<Args>(args)...);
                                                                                        ^
crow/routing.h:584:88: note: candidates are:
In file included from crow/routing.h:18:0,
                 from crow.h:19,
                 from main.cpp:1:
./crow/middleware.h:206:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:206:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:206:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:216:11: note: template<class F, class ... Args> typename std::enable_if<(crow::black_magic::is_callable<F, crow::request&, crow::response&, Args ...>::value && (! crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value))>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:216:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<(crow::black_magic::is_callable<F, crow::request&, crow::response&, Args ...>::value && (! crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value))>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:216:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:226:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:226:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:226:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:236:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::request, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:236:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::request, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:236:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:247:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, Args>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:247:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, Args>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:241:22:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:247:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
In file included from crow.h:19:0,
                 from main.cpp:1:
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
crow/routing.h:584:88: error: no matching function for call to ‘wrapped_handler_call(crow::request&, crow::response&, const crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51&, std::basic_string<char>)’
                   detail::wrapped_handler_call(req, res, f, std::forward<Args>(args)...);
                                                                                        ^
crow/routing.h:584:88: note: candidates are:
In file included from crow/routing.h:18:0,
                 from crow.h:19,
                 from main.cpp:1:
./crow/middleware.h:206:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:206:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:206:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:216:11: note: template<class F, class ... Args> typename std::enable_if<(crow::black_magic::is_callable<F, crow::request&, crow::response&, Args ...>::value && (! crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value))>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:216:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<(crow::black_magic::is_callable<F, crow::request&, crow::response&, Args ...>::value && (! crow::black_magic::is_callable<F, const crow::request, crow::response&, Args ...>::value))>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:216:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:226:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:226:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::response&, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:226:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:236:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::request, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:236:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, crow::request, Args ...>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:236:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
crow/routing.h: In instantiation of ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’:
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:247:11: note: template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, Args>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...)
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
           ^
./crow/middleware.h:247:11: note:   template argument deduction/substitution failed:
./crow/middleware.h: In substitution of ‘template<class F, class ... Args> typename std::enable_if<crow::black_magic::is_callable<F, Args>::value>::type crow::detail::wrapped_handler_call(crow::request&, crow::response&, const F&, Args&& ...) [with F = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
crow/routing.h:584:88:   required from ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:581:16:   required from ‘struct crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’
crow/routing.h:577:22:   required from ‘void crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’
crow/app.h:254:129:   required from ‘void crow::Crow<Middlewares>::validate() [with Middlewares = {}]’
crow/app.h:280:22:   required from ‘void crow::Crow<Middlewares>::run() [with Middlewares = {}]’
main.cpp:12:25:   required from here
./crow/middleware.h:247:11: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
In file included from /usr/include/boost/utility/compare_pointees.hpp:15:0,
                 from /usr/include/boost/optional/optional.hpp:55,
                 from /usr/include/boost/optional.hpp:15,
                 from crow/query_string.h:9,
                 from crow.h:2,
                 from main.cpp:1:
/usr/lib/gcc/x86_64-pc-linux-gnu/4.8.5/include/c++/functional:2334:2: error: ‘std::function<_Res(_ArgTypes ...)>::_Requires<std::__and_<std::__not_<std::is_same<typename std::decay<_Tp>::type, std::function<_Res(_ArgTypes ...)> > >, std::__or_<std::is_void<_Tp>, std::is_convertible<std::function<_Res(_ArgTypes ...)>::_Invoke<typename std::decay<_Tp>::type>, _Res> > >, std::function<_Res(_ArgTypes ...)>&> std::function<_Res(_ArgTypes ...)>::operator=(_Functor&&) [with _Functor = crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27; _Res = void; _ArgTypes = {crow::request&, crow::response&}; std::function<_Res(_ArgTypes ...)>::_Requires<std::__and_<std::__not_<std::is_same<typename std::decay<_Tp>::type, std::function<_Res(_ArgTypes ...)> > >, std::__or_<std::is_void<_Tp>, std::is_convertible<std::function<_Res(_ArgTypes ...)>::_Invoke<typename std::decay<_Tp>::type>, _Res> > >, std::function<_Res(_ArgTypes ...)>&> = std::function<void(crow::request&, crow::response&)>&]’, declared using local type ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = main()::__lambda53; Args = {}]::__lambda27’, is used but never defined [-fpermissive]
  operator=(_Functor&& __f)
  ^
/usr/lib/gcc/x86_64-pc-linux-gnu/4.8.5/include/c++/functional:2334:2: error: ‘std::function<_Res(_ArgTypes ...)>::_Requires<std::__and_<std::__not_<std::is_same<typename std::decay<_Tp>::type, std::function<_Res(_ArgTypes ...)> > >, std::__or_<std::is_void<_Tp>, std::is_convertible<std::function<_Res(_ArgTypes ...)>::_Invoke<typename std::decay<_Tp>::type>, _Res> > >, std::function<_Res(_ArgTypes ...)>&> std::function<_Res(_ArgTypes ...)>::operator=(_Functor&&) [with _Functor = crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27; _Res = void; _ArgTypes = {crow::request&, crow::response&, std::basic_string<char, std::char_traits<char>, std::allocator<char> >}; std::function<_Res(_ArgTypes ...)>::_Requires<std::__and_<std::__not_<std::is_same<typename std::decay<_Tp>::type, std::function<_Res(_ArgTypes ...)> > >, std::__or_<std::is_void<_Tp>, std::is_convertible<std::function<_Res(_ArgTypes ...)>::_Invoke<typename std::decay<_Tp>::type>, _Res> > >, std::function<_Res(_ArgTypes ...)>&> = std::function<void(crow::request&, crow::response&, std::basic_string<char>)>&]’, declared using local type ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda50; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’, is used but never defined [-fpermissive]
/usr/lib/gcc/x86_64-pc-linux-gnu/4.8.5/include/c++/functional:2334:2: error: ‘std::function<_Res(_ArgTypes ...)>::_Requires<std::__and_<std::__not_<std::is_same<typename std::decay<_Tp>::type, std::function<_Res(_ArgTypes ...)> > >, std::__or_<std::is_void<_Tp>, std::is_convertible<std::function<_Res(_ArgTypes ...)>::_Invoke<typename std::decay<_Tp>::type>, _Res> > >, std::function<_Res(_ArgTypes ...)>&> std::function<_Res(_ArgTypes ...)>::operator=(_Functor&&) [with _Functor = crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27; _Res = void; _ArgTypes = {crow::request&, crow::response&, std::basic_string<char, std::char_traits<char>, std::allocator<char> >}; std::function<_Res(_ArgTypes ...)>::_Requires<std::__and_<std::__not_<std::is_same<typename std::decay<_Tp>::type, std::function<_Res(_ArgTypes ...)> > >, std::__or_<std::is_void<_Tp>, std::is_convertible<std::function<_Res(_ArgTypes ...)>::_Invoke<typename std::decay<_Tp>::type>, _Res> > >, std::function<_Res(_ArgTypes ...)>&> = std::function<void(crow::request&, crow::response&, std::basic_string<char>)>&]’, declared using local type ‘crow::TaggedRule<Args>::operator()(Func&&) [with Func = crow::Crow<Middlewares>::validate() [with Middlewares = {}]::__lambda51; Args = {std::basic_string<char, std::char_traits<char>, std::allocator<char> >}]::__lambda27’, is used but never defined [-fpermissive]

```
@dranikpg would it be possible to have `wrapped_handler_call` compile under GCC 4.8 ? What do you think?

--- a/include/crow/common.h
+++ b/include/crow/common.h
@@ -325,7 +325,7 @@ constexpr crow::HTTPMethod method_from_string(const char* str)
                                                            throw std::runtime_error("invalid http method");
 }
 
-constexpr crow::HTTPMethod operator""_method(const char* str, size_t /*len*/) 
+constexpr crow::HTTPMethod operator"" _method(const char* str, size_t /*len*/)
 {
     return method_from_string( str );
 }

--- a/include/crow/middleware.h
+++ b/include/crow/middleware.h
@@ -200,39 +200,24 @@ namespace crow
             static const bool value = decltype(f<MW>(nullptr))::value;
         };
 
-        // wrapped_handler_call transparently wraps a handler call behind (req, res, args...)
         template<typename F, typename... Args>
-        typename std::enable_if<black_magic::is_callable<F, const crow::request, crow::response&, Args...>::value>::type
-          wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
-        {
-            static_assert(std::is_same<void, decltype(f(std::declval<crow::request>(), std::declval<crow::response&>(), std::declval<Args>()...))>::value,
-                          "Handler function with response argument should have void return type");
-
-            f(req, res, std::forward<Args>(args)...);
-        }
-
-        template<typename F, typename... Args>
-        typename std::enable_if<black_magic::is_callable<F, crow::request&, crow::response&, Args...>::value && !black_magic::is_callable<F, const crow::request, crow::response&, Args...>::value>::type
-          wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
-        {
-            static_assert(std::is_same<void, decltype(f(std::declval<crow::request&>(), std::declval<crow::response&>(), std::declval<Args>()...))>::value,
-                          "Handler function with response argument should have void return type");
-
-            f(req, res, std::forward<Args>(args)...);
-        }
-
-        template<typename F, typename... Args>
-        typename std::enable_if<black_magic::is_callable<F, crow::response&, Args...>::value>::type
+        //typename std::enable_if<black_magic::is_callable<F, Args...>::value>::type
+        typename std::enable_if<black_magic::CallHelper<F, black_magic::S<Args...>>::value, void>::type
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
         {
-            static_assert(std::is_same<void, decltype(f(std::declval<crow::response&>(), std::declval<Args>()...))>::value,
-                          "Handler function with response argument should have void return type");
+            static_assert(!std::is_same<void, decltype(f(std::declval<Args>()...))>::value,
+                          "Handler function cannot have void return type; valid return types: string, int, crow::response, crow::returnable");
 
-            f(res, std::forward<Args>(args)...);
+            res = crow::response(f(std::forward<Args>(args)...));
+            res.end();
         }
 
         template<typename F, typename... Args>
-        typename std::enable_if<black_magic::is_callable<F, crow::request, Args...>::value>::type
+        //typename std::enable_if<black_magic::is_callable<F, crow::request, Args...>::value>::type
+        typename std::enable_if<
+          !black_magic::CallHelper<F, black_magic::S<Args...>>::value &&
+            black_magic::CallHelper<F, black_magic::S<crow::request&, Args...>>::value,
+          void>::type
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
         {
             static_assert(!std::is_same<void, decltype(f(std::declval<crow::request>(), std::declval<Args>()...))>::value,
@@ -243,14 +228,51 @@ namespace crow
         }
 
         template<typename F, typename... Args>
-        typename std::enable_if<black_magic::is_callable<F, Args...>::value>::type
+        //typename std::enable_if<black_magic::is_callable<F, crow::response&, Args...>::value>::type
+        typename std::enable_if<
+          !black_magic::CallHelper<F, black_magic::S<Args...>>::value &&
+            !black_magic::CallHelper<F, black_magic::S<crow::request&, Args...>>::value &&
+            black_magic::CallHelper<F, black_magic::S<crow::response&, Args...>>::value,
+          void>::type
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
         {
-            static_assert(!std::is_same<void, decltype(f(std::declval<Args>()...))>::value,
-                          "Handler function cannot have void return type; valid return types: string, int, crow::response, crow::returnable");
+            static_assert(std::is_same<void, decltype(f(std::declval<crow::response&>(), std::declval<Args>()...))>::value,
+                          "Handler function with response argument should have void return type");
 
-            res = crow::response(f(std::forward<Args>(args)...));
-            res.end();
+            f(res, std::forward<Args>(args)...);
+        }
+
+        template<typename F, typename... Args>
+        //typename std::enable_if<black_magic::is_callable<F, crow::request&, crow::response&, Args...>::value && !black_magic::is_callable<F, const crow::request, crow::response&, Args...>::value>::type
+        typename std::enable_if<
+          !black_magic::CallHelper<F, black_magic::S<Args...>>::value &&
+            !black_magic::CallHelper<F, black_magic::S<crow::request&, Args...>>::value &&
+            !black_magic::CallHelper<F, black_magic::S<crow::response&, Args...>>::value &&
+          black_magic::CallHelper<F, black_magic::S<const crow::request&, crow::response&, Args...>>::value,
+          void>::type
+          wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
+        {
+            static_assert(std::is_same<void, decltype(f(std::declval<crow::request&>(), std::declval<crow::response&>(), std::declval<Args>()...))>::value,
+                          "Handler function with response argument should have void return type");
+
+            f(req, res, std::forward<Args>(args)...);
+        }
+
+        // wrapped_handler_call transparently wraps a handler call behind (req, res, args...)
+        template<typename F, typename... Args>
+        typename std::enable_if<
+          !black_magic::CallHelper<F, black_magic::S<Args...>>::value &&
+            !black_magic::CallHelper<F, black_magic::S<crow::request&, Args...>>::value &&
+            !black_magic::CallHelper<F, black_magic::S<crow::response&, Args...>>::value &&
+            !black_magic::CallHelper<F, black_magic::S<const crow::request&, crow::response&, Args...>>::value,
+          void>::type
+          //typename std::enable_if<black_magic::is_callable<F, const crow::request, crow::response&, Args...>::value>::type
+          wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
+        {
+            static_assert(std::is_same<void, decltype(f(std::declval<crow::request&>(), std::declval<crow::response&>(), std::declval<Args>()...))>::value,
+                          "Handler function with response argument should have void return type");
+
+            f(req, res, std::forward<Args>(args)...);
         }
 
         template<typename F, typename App, typename... Middlewares>

--- a/include/crow/middleware.h
+++ b/include/crow/middleware.h
@@ -248,7 +248,7 @@ namespace crow
           !black_magic::CallHelper<F, black_magic::S<Args...>>::value &&
             !black_magic::CallHelper<F, black_magic::S<crow::request&, Args...>>::value &&
             !black_magic::CallHelper<F, black_magic::S<crow::response&, Args...>>::value &&
-          black_magic::CallHelper<F, black_magic::S<const crow::request&, crow::response&, Args...>>::value,
+            black_magic::CallHelper<F, black_magic::S<const crow::request&, crow::response&, Args...>>::value,
           void>::type
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
         {

--- a/include/crow/middleware.h
+++ b/include/crow/middleware.h
@@ -201,7 +201,6 @@ namespace crow
         };
 
         template<typename F, typename... Args>
-        //typename std::enable_if<black_magic::is_callable<F, Args...>::value>::type
         typename std::enable_if<black_magic::CallHelper<F, black_magic::S<Args...>>::value, void>::type
           wrapped_handler_call(crow::request& /*req*/, crow::response& res, const F& f, Args&&... args)
         {
@@ -213,7 +212,6 @@ namespace crow
         }
 
         template<typename F, typename... Args>
-        //typename std::enable_if<black_magic::is_callable<F, crow::request, Args...>::value>::type
         typename std::enable_if<
           !black_magic::CallHelper<F, black_magic::S<Args...>>::value &&
             black_magic::CallHelper<F, black_magic::S<crow::request&, Args...>>::value,
@@ -228,7 +226,6 @@ namespace crow
         }
 
         template<typename F, typename... Args>
-        //typename std::enable_if<black_magic::is_callable<F, crow::response&, Args...>::value>::type
         typename std::enable_if<
           !black_magic::CallHelper<F, black_magic::S<Args...>>::value &&
             !black_magic::CallHelper<F, black_magic::S<crow::request&, Args...>>::value &&
@@ -243,7 +240,6 @@ namespace crow
         }
 
         template<typename F, typename... Args>
-        //typename std::enable_if<black_magic::is_callable<F, crow::request&, crow::response&, Args...>::value && !black_magic::is_callable<F, const crow::request, crow::response&, Args...>::value>::type
         typename std::enable_if<
           !black_magic::CallHelper<F, black_magic::S<Args...>>::value &&
             !black_magic::CallHelper<F, black_magic::S<crow::request&, Args...>>::value &&
@@ -266,7 +262,6 @@ namespace crow
             !black_magic::CallHelper<F, black_magic::S<crow::response&, Args...>>::value &&
             !black_magic::CallHelper<F, black_magic::S<const crow::request&, crow::response&, Args...>>::value,
           void>::type
-          //typename std::enable_if<black_magic::is_callable<F, const crow::request, crow::response&, Args...>::value>::type
           wrapped_handler_call(crow::request& req, crow::response& res, const F& f, Args&&... args)
         {
             static_assert(std::is_same<void, decltype(f(std::declval<crow::request&>(), std::declval<crow::response&>(), std::declval<Args>()...))>::value,

--- a/include/crow/utility.h
+++ b/include/crow/utility.h
@@ -233,6 +233,8 @@ namespace crow
             template<template<typename... Args> class U>
             using rebind = U<T...>;
         };
+
+        // Check whether the template function can be called with specific arguments
         template<typename F, typename Set>
         struct CallHelper;
         template<typename F, typename... Args>
@@ -262,19 +264,6 @@ namespace crow
         template<typename T, typename... Ts>
         struct has_type<T, std::tuple<T, Ts...>> : std::true_type
         {};
-
-        // Check F is callable with Args
-        template<typename F, typename... Args>
-        struct is_callable
-        {
-            template<typename F2, typename... Args2>
-            static std::true_type __test(decltype(std::declval<F2>()(std::declval<Args2>()...))*);
-
-            template<typename F2, typename... Args2>
-            static std::false_type __test(...);
-
-            static constexpr bool value = decltype(__test<F, Args...>(nullptr))::value;
-        };
 
         // Kind of fold expressions in C++11
         template<bool...>

--- a/tests/ssl/ssltest.cpp
+++ b/tests/ssl/ssltest.cpp
@@ -73,7 +73,7 @@ TEST_CASE("SSL")
         }
         else
         {
-            CHECK(std::string("Hello world, I'm keycrt.").substr((z*-1)) == to_test);
+            CHECK(std::string("Hello world, I'm keycrt.").substr((z * -1)) == to_test);
         }
     }
 

--- a/tests/ssl/ssltest.cpp
+++ b/tests/ssl/ssltest.cpp
@@ -55,6 +55,7 @@ TEST_CASE("SSL")
 
         size_t x = 0;
         size_t y = 0;
+        long z = 0;
 
         while (x < 121)
         {
@@ -63,7 +64,17 @@ TEST_CASE("SSL")
             buf[y] = '\0';
         }
 
-        CHECK(std::string("Hello world, I'm keycrt.") == std::string(buf));
+        std::string to_test(buf);
+
+        if ((z = to_test.length() - 24) >= 0)
+        {
+
+            CHECK(std::string("Hello world, I'm keycrt.") == to_test.substr(z));
+        }
+        else
+        {
+            CHECK(std::string("Hello world, I'm keycrt.").substr((z*-1)) == to_test);
+        }
     }
 
     /*

